crates/database/src/order_execution.rs

```diff
@@ -1,33 +1,91 @@
 use {
-    crate::{auction::AuctionId, OrderUid},
+    crate::{auction::AuctionId, Address, OrderUid},
     bigdecimal::BigDecimal,
-    sqlx::PgConnection,
+    sqlx::{
+        postgres::{PgHasArrayType, PgTypeInfo},
+        PgConnection,
+    },
+    std::collections::HashMap,
 };
 
+#[derive(Clone, Debug, Eq, PartialEq, sqlx::Type, sqlx::FromRow)]
+pub struct FeeAsset {
+    pub amount: BigDecimal,
+    pub token: Address,
+}
+
+// explains that the equivalent Postgres type is already defined in the database
+impl PgHasArrayType for FeeAsset {
+    fn array_type_info() -> PgTypeInfo {
+        PgTypeInfo::with_name("_feeasset")
+    }
+}
+
 pub async fn save(
     ex: &mut PgConnection,
     order: &OrderUid,
     auction: AuctionId,
     block_number: i64,
     executed_fee: &BigDecimal,
+    executed_protocol_fees: Option<&[FeeAsset]>,
 ) -> Result<(), sqlx::Error> {
     const QUERY: &str = r#"
-INSERT INTO order_execution (order_uid, auction_id, reward, surplus_fee, block_number)
-VALUES ($1, $2, $3, $4, $5)
+INSERT INTO order_execution (order_uid, auction_id, reward, surplus_fee, block_number, protocol_fees)
+VALUES ($1, $2, $3, $4, $5, $6)
 ON CONFLICT (order_uid, auction_id)
-DO UPDATE SET reward = $3, surplus_fee = $4, block_number = $5
+DO UPDATE SET reward = $3, surplus_fee = $4, block_number = $5, protocol_fees = $6
 ;"#;
     sqlx::query(QUERY)
         .bind(order)
         .bind(auction)
         .bind(0.) // reward is deprecated but saved for historical analysis
         .bind(Some(executed_fee))
         .bind(block_number)
+        .bind(executed_protocol_fees)
         .execute(ex)
         .await?;
     Ok(())
 }
 
+// fetch protocol fees for all keys in the filter
+pub async fn executed_protocol_fees(
+    ex: &mut PgConnection,
+    keys_filter: &[(AuctionId, OrderUid)],
+) -> Result<HashMap<(AuctionId, OrderUid), Vec<FeeAsset>>, sqlx::Error> {
+    if keys_filter.is_empty() {
+        return Ok(HashMap::new());
+    }
+
+    let mut fees = HashMap::new();
+    for (auction_id, order_uid) in keys_filter {
+        let protocol_fees = get_protocol_fees(ex, order_uid, *auction_id).await?;
+        fees.insert((*auction_id, *order_uid), protocol_fees.unwrap_or_default());
+    }
+
+    Ok(fees)
+}
+
+// executed procotol fee for a single <order, auction> pair
+async fn get_protocol_fees(
```

Comment:
isn't the name misleading if we aren't getting protocol fees but executed protocol fees?

Reply:
removed